tests/php/src/Internal/DependencyManagement/AbstractServiceProviderTests.php

```diff
@@ -0,0 +1,128 @@
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithDependencies;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\ClassWithPrivateConstructor;
+use Automattic\WooCommerce\Tests\Internal\DependencyManagement\ExampleClasses\DependencyClass;
+use League\Container\Definition\DefinitionInterface;
+
+/**
+ * Tests for AbstractServiceProvider.
+ */
+class AbstractServiceProviderTests extends \WC_Unit_Test_Case {
+
+	/**
+	 * The system under test.
+	 *
+	 * @var AbstractServiceProvider
+	 */
+	private $sut;
+
+	/**
+	 * The container used for tests.
+	 *
+	 * @var ExtendedContainer
+	 */
+	private $container;
+
+	/**
+	 * Runs before each test.
+	 */
+	public function setUp() {
+		$this->container = new ExtendedContainer();
+
+		$this->sut = new class() extends AbstractServiceProvider {
+			// phpcs:disable
+
+			/**
+			 * Public version of add_with_auto_arguments, which is usually protected.
+			 */
+			public function add_with_auto_arguments( string $class_name, $concrete = null, bool $shared = false ) : DefinitionInterface {
+				return parent::add_with_auto_arguments( $class_name, $concrete, $shared );
+			}
+
+			/**
+			 * The mandatory 'register' method (defined in the base class as abstract).
+			 * Not implemented because this class is tested on its own, not as a service provider actually registered on a container.
+			 */
+			public function register() {}
+
+			// phpcs:enable
+		};
+
+		$this->sut->setContainer( $this->container );
+	}
+
+	/**
+	 * @testdox 'add_with_auto_arguments' should throw an exception if an invalid class name is passed.
+	 *
+	 * @throws \Exception Invalid class name passed.
+	 */
+	public function test_add_with_auto_arguments_throws_on_non_class_passed() {
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( "AbstractServiceProvider::addWithAutoArguments: error when reflecting class 'foobar': Class foobar does not exist" );
+
+		$this->sut->add_with_auto_arguments( 'foobar' );
+	}
+
+	/**
+	 * @testdox 'add_with_auto_arguments' should throw an exception if the passed class has a private constructor.
+	 *
+	 * @throws \Exception The passed class has a private constructor.
+	 */
+	public function test_add_with_auto_arguments_throws_on_private_constructor() {
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( "AbstractServiceProvider::addWithAutoArguments: constructor of class '" . ClassWithPrivateConstructor::class . "' isn't public, instances can't be created." );
+
+		$this->sut->add_with_auto_arguments( ClassWithPrivateConstructor::class );
+	}
+
+	/**
+	 * @testdox 'add_with_auto_arguments' should throw an exception if the passed class has a constructor argument without type hint.
+	 *
+	 * @throws \Exception The passed class has a constructor argument without type hint.
```

Comment:
Documenting the specific exceptions thrown by testing functions seems redundant, what do you think?

Reply:
Yes but if we don't do that phpcs complaints 🤷🏻‍♂️

Reply:
I think it would be worth making broader exceptions (no pun intended 😅) for PHPCS rules within the `tests/` directory. How about disabling that sniff for tests?

Reply:
I tried again and phpcs isn't complaining about those anymore, not sure if we have changed some configuration recently. Anyway, I've removed the `@throw`s.